framework/aster-frame/src/task/processor.rs

```diff
@@ -63,7 +67,7 @@ pub fn schedule() {
 
 pub fn preempt() {
     // disable interrupts to avoid nested preemption.
-    let disable_irq = disable_local();
+    disable_irq_if_enabled();
```

Comment:
If the function returns early, the IRQs are disabled forever, which is not correct.

For example, if the function exits here:
```rust
    if !scheduler.should_preempt(&curr_task) {
        return;
    }
```

I also wonder what made you drop your original solution of moving the preempt check from the beginning of `switch_to_task()` to the beginning of `preempt()` and `schedule()`? That solution sounds a bit more correct than the one proposed in this PR. Does it have any problems I'm not aware of?

Reply:
> I also wonder what made you drop your original solution of moving the preempt check from the beginning of switch_to_task() to the beginning of preempt() and schedule()? That solution sounds a bit more correct than the one proposed in this PR. Does it have any problems I'm not aware of?

OK, I think I know why. We should not really have a non-zero preempt count when we call `switch_task', otherwise the preempt count (as a CPU-local variable) will remain non-zero even after the switch has finished, which is obviously not correct.

So now we have two guards, the preempt guard and the preempt&IRQ guard. The question is whether we need the third guard, the IRQ guard, and if so, we can manually make the preempt&IRQ guard from the preempt guard and the IRQ guard.